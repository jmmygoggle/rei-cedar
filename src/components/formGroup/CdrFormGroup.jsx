import clsx from 'clsx';
import style from './styles/CdrFormGroup.scss';
import IconErrorStroke from '../icon/comps/error-stroke';
import CdrFormError from '../formError/CdrFormError';

export default {
  name: 'CdrFormGroup',
  components: {
    IconErrorStroke,
    CdrFormError,
  },
  props: {

    id: String,
    label: {
      type: String,
      default: '',
      required: false,
    },
    // Set error styling
    error: {
      type: [Boolean, String],
      default: false,
    },
    /**
    * Override the error message role, default is `status`.
    */
    errorRole: {
      type: String,
      required: false,
      default: 'status',
    },
    required: Boolean,
    optional: Boolean,
    disabled: Boolean,
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-form-group';
    },
    errorClass() {
      return this.error ? this.style['cdr-form-group--error'] : '';
    },
    disabledClass() {
      return this.disabled ? this.style['cdr-form-group--disabled'] : '';
    },
    groupId() {
      return this.id ? this.id : this._uid; // eslint-disable-line no-underscore-dangle
    },
  },
  render() {
    const requiredEl = this.required ? (
      <span aria-label="required">
        *
      </span>
    ) : '';

    const optionalEl = this.optional ? (
      <span
        class={this.style['cdr-form-group__optional']}
      >
        (optional)
      </span>
    ) : '';

    return (
      <fieldset
        class={clsx(this.style[this.baseClass], this.disabledClass)}
        disabled={this.disabled}
        aria-invalid={!!this.error}
        aria-errormessage={!!this.error && `${this.groupId}-error`}
      >
        <legend>
          {this.$slots.label || this.label}
          {requiredEl || optionalEl ? ' ' : ''}
          {requiredEl || optionalEl}
        </legend>
        <div class={clsx(this.style['cdr-form-group__wrapper'], this.errorClass)}>
          {this.$slots.default}
        </div>
        {this.error && (
          <cdr-form-error role={this.errorRole} error={this.error} id={`${this.groupId}-error`}>
            <template slot="error">
              {this.$slots.error}
            </template>
          </cdr-form-error>
        )}
      </fieldset>
    );
  },
};
