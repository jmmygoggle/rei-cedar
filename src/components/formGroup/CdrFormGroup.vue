<template>
  <fieldset
    :class="[$style[baseClass], $style[disabledClass]]"
    :disabled="disabled"
  >
    <legend>
      <slot name="label" /><span v-if="!hasLabelSlot">{{ label }}</span>
      <span v-if="required || optional" />
      <span aria-label="required">*</span>
      <span :class="$style['cdr-form-group__optional']">
        (optional)
      </span>
    </legend>
    <div :class="[$style['cdr-form-group__wrapper'], $style[errorClass]]">
      <slot />
    </div>
    <cdr-form-error
      :error="error"
      v-if="error"
    >
      <template slot="error">
        <slot name="error" />
      </template>
    </cdr-form-error>
  </fieldset>
</template>
<script lang="ts">
import { defineComponent, computed } from 'vue';
import CdrFormError from '../formError/CdrFormError';

export default defineComponent({
  name: 'CdrFormGroup',
  components: {
    CdrFormError,
  },
  props: {
    label: {
      type: String,
      default: '',
      required: false,
    },
    // Set error styling
    error: {
      type: [Boolean, String],
      default: false,
    },
    required: Boolean,
    optional: Boolean,
    disabled: Boolean,
  },
  setup(props, ctx) {
    const baseClass = 'cdr-form-group';
    const errorClass = computed(() => props.error && 'cdr-form-group--error');
    const disabledClass = computed(() => props.disabled && 'cdr-form-group--disabled');
    const hasLabelSlot = ctx.slots.label;
    return {
      baseClass,
      errorClass,
      disabledClass,
      hasLabelSlot,
    };
  },
});
</script>

<style lang="scss" module src="./styles/CdrFormGroup.scss">
</style>
