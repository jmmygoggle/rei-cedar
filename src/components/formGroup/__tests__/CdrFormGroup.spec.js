import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrFormGroup from 'componentdir/formGroup/CdrFormGroup';

describe('CdrFormGroup', () => {
  test('renders correctly', () => {
    const wrapper = mount(CdrFormGroup, {
      propsData: {
        label: 'hey'
      },
      slots: {
        'default': 'form elements!',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('renders error slot', () => {
    const wrapper = mount(CdrFormGroup, {
      propsData: {
        error: true
      },
      slots: {
        'error': 'whoops',
      },
    });

    expect(wrapper.find('.cdr-form-group--error').exists()).toBe(true);
    expect(wrapper.find('.cdr-form-error').text()).toBe('whoops');
  });

  test('renders disabled state', () => {
    const wrapper = mount(CdrFormGroup, {
      propsData: {
        disabled: true
      },
    });

    expect(wrapper.find('.cdr-form-group--disabled').exists()).toBe(true);
  });

  test('renders text when passed as error prop', () => {
    const wrapper = mount(CdrFormGroup, {
      propsData: {
        error: 'false!'
      },
    });

    expect(wrapper.find('.cdr-form-group--error').exists()).toBe(true);
    expect(wrapper.find('.cdr-form-error').text()).toBe('false!');
  });

  test('does not render error slot if error state is inactive', () => {
    const wrapper = mount(CdrFormGroup, {
      propsData: {
        error: false
      },
      slots: {
        'error': 'whoops',
      },
    });
    expect(wrapper.find('.cdr-form-group--error').exists()).toBe(false);
    expect(wrapper.find('.cdr-form-error').exists()).toBe(false);
  });
});
