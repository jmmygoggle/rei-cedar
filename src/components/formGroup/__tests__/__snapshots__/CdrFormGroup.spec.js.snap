// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrFormGroup renders correctly 1`] = `
<fieldset
  class="cdr-form-group"
>
  <legend>
    
    hey
    
    <!--v-if-->
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper"
  >
    
    form elements!
    
  </div>
  <!--v-if-->
</fieldset>
`;
